src/script/backup/BackUpHeader.tsx

```diff
@@ -0,0 +1,210 @@
+    } = this;
+
+    const salt = sodium.randombytes_buf(sodium.crypto_pwhash_SALTBYTES);
+    const hashedUserId = this.hashUserId(
+      this.userId,
+      salt,
+      this.OPSLIMIT_INTERACTIVE_VALUE,
+      this.MEMLIMIT_INTERACTIVE_VALUE,
+    );
+    const formatBytes = sodium.from_string(this.format);
+    const versionBytes = sodium.from_string(this.version);
+    const nonReadableByte = new Uint8Array([0x00]);
+    const opslimitBytes = new Uint8Array(UNSIGNED_INT_LENGTH);
+    const memlimitBytes = new Uint8Array(UNSIGNED_INT_LENGTH);
+
+    // Set opslimitBytes and memlimitBytes using DataView
+    new DataView(opslimitBytes.buffer).setUint32(0, this.OPSLIMIT_INTERACTIVE_VALUE, false);
+    new DataView(memlimitBytes.buffer).setUint32(0, this.MEMLIMIT_INTERACTIVE_VALUE, false);
+
+    const headerData = new Uint8Array(
+      BACKUP_HEADER_FORMAT_LENGTH +
+        BACKUP_HEADER_EXTRA_GAP_LENGTH +
+        BACKUP_HEADER_VERSION_LENGTH +
+        sodium.crypto_pwhash_SALTBYTES +
+        PWD_HASH_OUTPUT_BYTES +
+        2 * UNSIGNED_INT_LENGTH,
+    );
+    let offset = 0;
+    offset = this.copyBytes(headerData, formatBytes, offset);
+    offset = this.copyBytes(headerData, nonReadableByte, offset);
+    offset = this.copyBytes(headerData, versionBytes, offset);
+    offset = this.copyBytes(headerData, salt, offset);
+    offset = this.copyBytes(headerData, hashedUserId, offset);
+    offset = this.copyBytes(headerData, opslimitBytes, offset);
+    this.copyBytes(headerData, memlimitBytes, offset);
+
+    return headerData;
+  }
+
+  copyBytes(destination: Uint8Array, source: Uint8Array, offset: number): number {
+    destination.set(source, offset);
+    const nextPos = offset + source.length;
+    return nextPos;
+  }
+
+  async decodeHeader(
+    encryptedDataSource: Uint8Array,
+  ): Promise<{decodingError: string | null; decodedHeader: DecodedHeader; headerSize: number}> {
+    const dataSrc = new Uint8Array(encryptedDataSource);
+    const {decodedHeader, headerSize} = this.readBackupHeader(dataSrc);
+
+    await sodium.ready;
+    // Sanity checks
+    const expectedHashedUserId = this.hashUserId(
+      this.userId,
+      decodedHeader.salt,
+      decodedHeader.opslimit,
+      decodedHeader.memlimit,
+    );
+    const storedHashedUserId = decodedHeader.hashedUserId;
+    const decodingError = this.handleHeaderDecodingErrors(decodedHeader, expectedHashedUserId, storedHashedUserId);
+
+    return {decodingError, decodedHeader, headerSize};
+  }
+
+  handleHeaderDecodingErrors(
+    decodedHeader: DecodedHeader,
+    expectedHashedUserId: Uint8Array,
+    storedHashedUserId: Uint8Array,
+  ) {
+    const {format, version} = decodedHeader;
+    if (!sodium.memcmp(expectedHashedUserId, storedHashedUserId)) {
+      return 'INVALID_USER_ID';
+    } else if (format !== this.format) {
+      return 'INVALID_FORMAT';
+    } else if (parseInt(version) < parseInt(this.version)) {
+      return 'INVALID_VERSION';
```

Comment:
Let's have a const / enum for validation cases ('INVALID_USER_ID' and the other 2)